### PR TITLE
Proposal: Preferences API

### PR DESCRIPTION
**Pull request by mickael-menu**
[See the full proposal](https://github.com/readium/architecture/blob/proposal/presentation-api/proposals/000-preferences-api.md)

## Summary

The Preferences API is a tool for creating components configurable during runtime. It provides a useful framework for constructing user settings interfaces, taking into account all the rules and dependencies associated with each setting. This makes it easy for developers to create a user interface that can be customized to their individual needs.

Multiple Readium components, such as Navigators, use this new framework to provide a unified API that enables them to change their settings on-the-fly.

## Motivation

Navigators in Readium are built differently depending on the type of publication. In the past, the API for modifying their presentation varied across implementations, making it more difficult to build a user interface and store user preferences.

Developers now have access to a single API enabling them to reuse user interface views across types of publications, such as:

* Visual publication with adjustable fonts and dimensions, e.g. reflowable EPUB, HTML document or PDF with reflow mode enabled.
* Visual publication with a fixed layout, e.g. fixed-layout EPUB, PDF or comic books.
* Publication with media playback, e.g. audiobook, text-to-speech or EPUB media overlays settings.

### CODE REVIEW

**Comment by stevenzeck** [comment]
I think for the most part this is good.  But a few things:

1. I'm still not 100% clear what the difference is between `PresentationProperties` and `PresentationSettings`.  If that could be made clearer under the Terms section, that would be helpful.  Maybe even a class diagram or something
2. I haven't really looked at the Swift implementation, but is it using [Key-Value Observing](https://developer.apple.com/documentation/combine/performing-key-value-observing-with-combine) at all?
3. I'm not sure about the name `SwitchProperty`.  Using "Switch" feels more like a user interface term.  Other suggestions:
    1. ToggleProperty
    2. BooleanProperty
    3. BinaryProperty

**Comment by mickael-menu** [comment]
@stevenzeck 

> 1. I'm still not 100% clear what the difference is between `PresentationProperties` and `PresentationSettings`.  If that could be made clearer under the Terms section, that would be helpful.  Maybe even a class diagram or something

I agree that it is a bit confusing and I'm not 100% happy about this part. But I'm not sure how to make it clearer in the Terms section, maybe it's a sign that this part is not properly designed or named.

Basically I separated two things:

* **Properties** which are dynamic values representing the current presentation state of the Navigator. The navigator also provides metadata about each property (ranges, constraints, localized description, etc.).
* **Settings** are a bunch of values that the app would like the Navigator to use when computing the **Properties** values. The Navigator still decides what is the actual value and might ignore some of the settings, so it's not a "hard-set".

To associate the properties and settings, I added the concept of a **Property Key** which is just a string identifier, e.g. `fontSize` or `com.company.customKey`.

> 2. I haven't really looked at the Swift implementation, but is it using [Key-Value Observing](https://developer.apple.com/documentation/combine/performing-key-value-observing-with-combine) at all?

There's no implementation yet but KVO would be a good fit yes, I think. I was thinking of Combine but it's only available on iOS 13+ so not usable in Readium.

> 3. I'm not sure about the name `SwitchProperty`.  Using "Switch" feels more like a user interface term.  Other suggestions:
>    
>    1. ToggleProperty
>    2. BooleanProperty
>    3. BinaryProperty

I like `ToggleProperty` as well.

**Comment by HadrienGardeur** [comment]
To avoid the confusion between Presentation Settings and Properties, maybe we could just say that a list of Property Keys is passed to Presentation Properties ?

This way there's no need to add another concept term.

**Comment by mickael-menu** [comment]
How about renaming `PresentationSettings` into `PresentationKeys`, keeping the same capabilities? This way we remove the concept of "Presentation Settings" and we just get a map of Presentation Key -> Value wrapped in a `PresentationKeys` object for convenience (type-safe accessors).

File: proposals/000-settings-api.md [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+
+#### Properties
+
+* `range: Range<V>`
+    * The valid range for the setting value.
+* `suggestedSteps: List<V>?`
+    * Value steps which can be used to decrement or increment the setting. It **must** be sorted in increasing order.
+* `suggestedIncrement: V?`
+    * Suggested value increment which can be used to decrement or increment the setting.
+* `label: (V) -> String`
```
Comment (qnga) [inline proposals/000-settings-api.md:318]: I would name this differently. At first glance, I was under the impression that it was a label for the setting. What about `formattedValue` or `prettyValue` ?

Comment (mickael-menu) [reply proposals/000-settings-api.md:318]: I agree, since it's a function I would name it `formatValue: (V) -> String`.

**Pull request merged**